File: 46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        do {
            res.push_back(vector<int>(nums.begin(), nums.end()));
        } while (nextPermutation(nums));
        return res;
    }
    
    bool nextPermutation(vector<int>& nums) {
        if (nums.size() <= 1) return false;
        
        int p1 = -1, p2;
        for (int i = nums.size()-2; i >= 0; i--)
            if (nums[i] < nums[i+1]) {
                p1 = i;
                break;
            }
            
        if (p1 == -1) return false;
        
        p2 = p1+1;
        // mind that nums[i] > nums[p1] && nums[i] <= nums[p2]
        for (int i = p2; i < nums.size(); i++)
            if (nums[i] > nums[p1] && nums[i] <= nums[p2])
                p2 = i;
        
        swap(nums[p1], nums[p2]);
        reverse(nums.begin()+p1+1, nums.end());
        return true;
    }
};

File: 46.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<>();
        for (int num : nums) path.add(num);
        
        generate(result, path, nums.length, 0);
        return result;
    }
    
    private void generate(
        List<List<Integer>> result,
        List<Integer> path,
        int n,
        int start
    ) {
        if (start == n) {
            result.add(new ArrayList<Integer>(path));
            return;
        }
        
        for (int i = start; i < n; i++) {
            swap(path, start, i);
            generate(result, path, n, start+1);
            swap(path, start, i);
        }
    }
    
    private void swap(List<Integer> list, int l, int r) {
        int t = list.get(l);
        list.set(l, list.get(r));
        list.set(r, t);
    }
}
